cfpb/cfgov-refresh
# New Wagtail page type for Job Listings

This PR implements a new Wagtail page type and associated panel types to allow for creation and editing of job listings through the Wagtail admin site.

Job listings follow this data model (all fields are required unless otherwise specified):

- Title
- Description
- Open/close dates
- Salary min/max
- Division
- Grade(s), one or more
- Region(s), one or more
- Application link(s), one or more of these two types:
  - USAJobs link, consisting of:
    - USAJobs announcement number
    - USAJobs URL
    - USAJobs applicant type
  - Email link, consisting of:
    - Email address
    - Email label (heading)
    - Email description (optional)


## Additions

- New Wagtail page type `jobmanager.models.pages.JobListingPage` for a single job listing.
- New Panel types for the various related models that need to be part of a job listing:
  - `jobmanager.models.panels.EmailApplicationLink` for email application links.
  - `jobmanager.models.panels.USAJobsApplicationLink` for USAJOBS application links.
  - `jobmanager.models.panels.GradePanel` for job grades.
  - `jobmanager.modes.panels.RegionPanel` for job regions.
- New section of the Wagtail admin for creating and editing all data related to job listings (see screenshots below).

## Changes

- Added `wagtail.contrib.modeladmin` to the list of `INSTALLED_APPS` to be able to use Wagtail's [ModelAdmin](http://docs.wagtail.io/en/v1.5.3/reference/contrib/modeladmin.html).
- Various edits to the job listing page template created in #2296 to support data model.
- Refactored `jobmanager.models` as a package with submodules as opposed to a single module file. Moved "legacy" Django models to `jobmanager.models.django`.
- ~~Change to page sharing logic to support use of ForeignKeys in Wagtail pages. See comment below for additional detail; this should be reviewed.~~ (Reverted this after modifying how some of the foreign key models were referenced and it no longer seems to be needed.)

## Testing

- Run `tox` or `tox -e fast` to verify that all unit tests pass, including new ones.

These work best if you have job data already loaded in your DB, ideally from a production data dump of the `jobmanager` models as a fixture, then loaded via `manage.py loaddata jobs.json`:

- Explore the various job-related pieces of data in the Wagtail admin:
 1. In the Wagtail admin, look at the "Job listings" menu in the vertical nav bar.
 2. Select, for example, "Regions", and note that you can create or edit relevant job-related data.
- Using the Wagtail admin, create a new job listing page locally:
 1. Run `./initial-data` to make sure that you have a root page and all migrations are applied.
 2. Login to the Wagtail admin at [http://localhost:8000/admin/](http://localhost:8000/admin/).
 3. Select "Pages", then under "CFGOV", choose "Add Child Page", "Job Listing Page" to start editing a new page.
 4. Enter a title. Select a divison. Add one or more grades. Add one or more regions. Select an open date and a close date. Enter a minimum and maximum salary. Enter a description. Add one ore more USAJOBS application links and/or email application links.
 5. Select "Preview" to see the page previewed in a new tab.
 6. Select "Save draft" to save the page as draft.

## Review

- @cfpb/cfgov-backends @schaferjh @sebworks @kurtrwall 

## Screenshots

New job listings menu in the Wagtail admin:

![image](https://cloud.githubusercontent.com/assets/654645/17679812/23fb27a4-630b-11e6-9461-f42de29b32e6.png)

Create a new Job Listing page:

![image](https://cloud.githubusercontent.com/assets/654645/17679822/34e787d8-630b-11e6-9845-14bdad3ea0df.png)

Job Listing page editor:

![image](https://cloud.githubusercontent.com/assets/654645/17679837/4e07430c-630b-11e6-8ba5-ec5e69da4faf.png)

Filling in job details:

![image](https://cloud.githubusercontent.com/assets/654645/17679854/670131c4-630b-11e6-9adf-a8d1e925dda8.png)

Filling in the application links:

![image](https://cloud.githubusercontent.com/assets/654645/17679883/8f89077a-630b-11e6-91f1-330f1c8ee819.png)

How application links appear:

![image](https://cloud.githubusercontent.com/assets/654645/17679918/ad8f7c72-630b-11e6-861a-08f7f8dc63b4.png)

Editing job-related data in the Wagtail admin:

![image](https://cloud.githubusercontent.com/assets/654645/17679940/c1c41b3a-630b-11e6-9913-97bde04472ae.png)

## Notes

- This PR replaces existing defunct #2298 as using the ModelAdmin approach seems more flexible and generally a better fit than using Snippets. This approach retains the existing models in `jobmanager.models.django` but adds Wagtail panel types on top of them to allow them to be selected in the Wagtail admin.
- When reviewing this PR, if there are minor improvements or changes that don't address the general approach being used here, please indicate if your suggestions are blockers, or if they can be addressed in a separate PR. There will likely need to be additional rounds of iteration on this but it would be nice to get this merged in as a start.

## Todos

- This PR purposefully does not address migration of existing job data from existing Django tables.
- Currently the page creation/edit form does not enforce that the user select one or more grades, regions, or application links. I looked into this a bit but it's tricky because it has to be done as part of page validation. This should be addressed in another PR.
- The sidebar needs to be properly filled in with content on the job listing page.

## Checklist
